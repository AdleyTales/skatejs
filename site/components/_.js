/** @jsx h */

import val from "@skatejs/val";
import { Component, define, h as preactH, props } from "../../src";

export const h = val(preactH);

function format(src) {
  // Format HTML so it displays.
  src = src.replace(/</g, "&lt;");
  src = src.replace(/>/g, "&gt;");

  // Remove leading newlines and only allow up to two newlines in code.
  src = src.split("\n").filter((v, i, a) => a[i - 1] || v.trim().length);

  // Get the initial indent so we can remove it from subsequent lines.
  const indent = src[0].match(/^\s*/)[0].length;

  // Format indentation.
  src = src.map(s => s.substring(indent));

  // Re-instate newline formatting.
  return src.join("\n");
}

export const Code = define(
  class extends Component {
    static props = {
      src: props.string
    };
    renderCallback({ src }) {
      return (
        <pre>
          <style>{`
            :host { display: block }
          `}</style>
          <code>
            {format(src)}
          </code>
        </pre>
      );
    }
  }
);

export const Heading = ({ children }) =>
  <h1>
    {children}
  </h1>;

export const Hero = define(
  class extends Component {
    renderCallback() {
      return (
        <div>
          <style>{`

          `}</style>
          <slot />
        </div>
      );
    }
  }
);

export const Layout = define(
  class extends Component {
    renderCallback() {
      return <slot />;
    }
  }
);
