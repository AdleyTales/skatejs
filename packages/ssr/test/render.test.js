require('./_');

const render = require('..');

function dom(html) {
  const div = document.createElement('div');
  div.innerHTML = html;
  return div.firstElementChild;
}

class Hello extends HTMLElement {
  connectedCallback() {
    this.attachShadow({ mode: 'open' }).appendChild(
      dom(`
        <span>
          Hello,
          <x-yell>
            <slot></slot>
          </x-yell>!
          <x-yell>!</x-yell>
        </span>
      `)
    );
  }
}
class Yell extends HTMLElement {
  connectedCallback() {
    Promise.resolve().then(() => {
      this.attachShadow({ mode: 'open' }).innerHTML = `
        <style>.test1,.test2 { font-weight: bold; }</style>
        <span class="test1"><slot></slot></span>
      `;
    });
  }
}
customElements.define('x-hello', Hello);
customElements.define('x-yell', Yell);

test('serialization: should call the connected callback', () => {
  const node = new CustomElement();
  return render(node).then(() => {
    expect(node.connectedCallback.mock.calls.length).toEqual(1);
  });
});

test('serialization: should disconnect it after serialising', () => {
  const node = new CustomElement();
  return render(node).then(() => {
    expect(node.disconnectedCallback.mock.calls.length).toEqual(1);
  });
});

test('rendering: node', () => {
  const hello = new Hello();
  hello.appendChild(document.createTextNode('World'));
  return render(hello).then(r => {
    expect(r).toMatchSnapshot();
  });
});

test('rendering: document', () => {
  const hello = new Hello();
  hello.appendChild(document.createTextNode('World'));
  document.body.innerHTML = '';
  document.head.innerHTML = '';
  document.body.appendChild(hello);
  return render(document).then(r => {
    document.body.removeChild(hello);
    expect(r).toMatchSnapshot();
  });
});
