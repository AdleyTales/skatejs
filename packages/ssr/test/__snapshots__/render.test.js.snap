// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render (serialisation) renders correctly document 1`] = `
"<!doctype html><html><head></head><body><script>function __ssr(){var s=document.currentScript,r=s.previousElementSibling,h=r.parentNode;h.removeChild(s);h.removeChild(r);r.parentNode.attachShadow({mode:f.getAttribute('mode')||'open'}).innerHTML=r.innerHTML;}</script><x-hello><shadowroot><span>Hello, <x-yell><shadowroot>
            <style>.test1,.test2 { font-weight: bold; }</style>
            <span class=\\"test1\\"><slot></slot></span>
          </shadowroot><script>__ssr()</script><slot></slot></x-yell>!<x-yell><shadowroot>
            <style>.test1,.test2 { font-weight: bold; }</style>
            <span class=\\"test1\\"><slot></slot></span>
          </shadowroot><script>__ssr()</script>!</x-yell></span></shadowroot><script>__ssr()</script>World</x-hello></body></html>"
`;

exports[`render (serialisation) renders correctly node 1`] = `
"<x-hello><shadowroot><span>Hello, <x-yell><shadowroot>
            <style>.test1,.test2 { font-weight: bold; }</style>
            <span class=\\"test1\\"><slot></slot></span>
          </shadowroot><script>__ssr()</script><slot></slot></x-yell>!<x-yell><shadowroot>
            <style>.test1,.test2 { font-weight: bold; }</style>
            <span class=\\"test1\\"><slot></slot></span>
          </shadowroot><script>__ssr()</script>!</x-yell></span></shadowroot><script>__ssr()</script>World</x-hello>"
`;
